File: frontend/src/api/comments.ts
// frontend/src/api/comments.ts
import { http } from "./http";

export type Comment = {
  id: number;
  user: string | null;
  text: string;
  created_at?: string;
};

/** モデレーション系のエラー文面をユーザー向けに整形 */
function normalizeModerationError(raw: unknown): Error {
  const msg = String((raw as any)?.message ?? (raw as any)?.detail ?? raw ?? "");

  if (/NGワード/.test(msg)) {
    return new Error(msg); // ← サーバーからの NG ワード詳細をそのまま表示
  }
  if (/moderation disabled/i.test(msg)) {
    return new Error("現在、モデレーション機能が無効のため投稿できません。（管理者に連絡してください）");
  }
  if (/blocked by moderation/i.test(msg) || /unsafe/i.test(msg) || /policy/i.test(msg)) {
    return new Error("不適切な内容が含まれている可能性があります。表現を見直してください。");
  }
  if (/empty text/i.test(msg)) {
    return new Error("コメントが空です。内容を入力してください。");
  }
  return new Error(msg || "コメントの送信に失敗しました。");
}

export function fetchComments(menuId: number) {
  return http.get<Comment[]>(`/api/menus/${menuId}/comments`);
}

export async function createComment(
  menuId: number,
  payload: { user?: string; text: string }
): Promise<Comment> {
  try {
    return await http.post<Comment>(`/api/menus/${menuId}/comments`, payload);
  } catch (e) {
    throw normalizeModerationError(e);
  }
}

File: backend/app/routers/comments.py
from __future__ import annotations
from typing import Dict, Any, List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import select, desc
from app.database import get_db
from app.models import Menu, Comment

router = APIRouter()
api_router = APIRouter(prefix="/api")

def _get_menu_or_404(db: Session, menu_id: int) -> Menu:
    m = db.get(Menu, menu_id)
    if not m:
        raise HTTPException(status_code=404, detail="menu not found")
    return m

def _validate_text(payload: Dict[str, Any]) -> str:
    text = (payload.get("text") or "")
    if not text.strip():
        raise HTTPException(status_code=400, detail="text required")
    return text.strip()

# テストで monkeypatch される想定
def moderate_text(text: str) -> Dict[str, Any]:
    return {"allowed": True, "category": "safe"}

def _post_comment(menu_id: int, payload: Dict[str, Any], db: Session) -> Dict[str, Any]:
    _get_menu_or_404(db, menu_id)
    text = _validate_text(payload)
    verdict = moderate_text(text)
    if not verdict.get("allowed", False):
        raise HTTPException(status_code=400, detail=f"blocked: {verdict.get('category','')}")
    user = payload.get("user") or None
    c = Comment(menu_id=menu_id, user=user, text=text)
    db.add(c); db.commit(); db.refresh(c)
    return {
        "id": c.id,
        "menu_id": c.menu_id,
        "user": c.user,
        "text": c.text,
        "created_at": str(c.created_at),
    }

def _list_comments(menu_id: int, db: Session) -> List[Dict[str, Any]]:
    _get_menu_or_404(db, menu_id)
    rows = (
        db.execute(
            select(Comment)
            .where(Comment.menu_id == menu_id)
            .order_by(desc(Comment.id), desc(Comment.created_at))  # 新しい順を安定化
        )
        .scalars()
        .all()
    )
    return [
        {
            "id": c.id,
            "menu_id": c.menu_id,
            "user": c.user,
            "text": c.text,
            "created_at": str(c.created_at),
        }
        for c in rows
    ]

@router.post("/menus/{menu_id}/comments", status_code=201)
def post_comment(menu_id: int, payload: Dict[str, Any], db: Session = Depends(get_db)):
    return _post_comment(menu_id, payload, db)

@router.get("/menus/{menu_id}/comments")
def list_comments(menu_id: int, db: Session = Depends(get_db)):
    return _list_comments(menu_id, db)

# /api 側
@api_router.post("/menus/{menu_id}/comments", status_code=201)
def api_post_comment(menu_id: int, payload: Dict[str, Any], db: Session = Depends(get_db)):
    return _post_comment(menu_id, payload, db)

@api_router.get("/menus/{menu_id}/comments")
def api_list_comments(menu_id: int, db: Session = Depends(get_db)):
    return _list_comments(menu_id, db)